source/agora/consensus/EnrollmentManager.d

```diff
@@ -682,6 +642,62 @@ public class EnrollmentManager
         }
     }
 
+    /***************************************************************************
+
+        Generate pre-images
+
+        Params:
+            current_height = block height used to determine which range of
+                pre-images will be generated.
+
+    ***************************************************************************/
+
+    private void generatePreimages(ulong current_height) @safe nothrow
+    {
+        // This determines which range of preimages must be generated.
+        ulong start_height =
+            ((current_height / ValidatorSet.ValidatorCycle) + 2) *
+                ValidatorSet.ValidatorCycle;
+        ulong bulk_index = start_height / AllCountPrimages;
+        const ulong MaxHeight = (bulk_index + 1) * AllCountPrimages;
```

Comment:
I have a hard time understanding this calculation here? Why the `+ 2` for example? Doesn't that mean that `start_height` will always be >= 2016 ?

Reply:
Considering your question, the following lines could be more accurate.
```
if ((current_height - 1) / ValidatorSet.ValidatorCycle == 0) 
    start_height = ((current_height / ValidatorSet.ValidatorCycle) + 1) * ValidatorSet.ValidatorCycle;
else
     start_height = ((current_height / ValidatorSet.ValidatorCycle) + 2) * ValidatorSet.ValidatorCycle;
```
The code is just for explaining.
Based on the above code, we will get 1008, 2016, 3124, ... for `start_height`.

For detailed example,  if the start height is 1800, it needs the 3024th hash value of the this.preimages. For detail, 1800 / 1008 is 1, and the value + 2 is 3, and 3 * 1008 is 3024.

But, the original code is not bad because we don't know how many preimages need to be prepared. It's because the height at which a node requests for a registration and the enrolled height are different.

Reply:
Why not just:
```D
start_height = current_height + (ValidatorSet.ValidatorCycle - (current_height % ValidatorSet.ValidatorCycle));
```

Reply:
I will explain an example based on the code suggested by you. If `current_height` is 1800, `start_height` will be 2016. So we will be able to get only 217 random seeds from the height of 1800 to 2016 but we need more than 1008 random seeds.

Reply:
How about adding `+ ValidatorSet.ValidatorCycle` then ?

Reply:
If my code and your code output same result, I'd like to choose my code because my code is simpler.

Reply:
Alright, but then you need to handle `current_height == 0` and please add a comment why we have `+ 1` and `+ 2`.

Reply:
> Alright, but then you need to handle current_height == 0 and please add a comment why we have + 1 and + 2.

Sure, thank you.